# Feedback from previous weeks and other and hints

1. Be careful about your environment remembering variables. Make sure your code works in a new _clean_ environment. In Colab: `Runtime`->`restart Runtime`, in Anaconda's Jupyter: `Kernel`->`Restart`.
2. Graphs without labels (or units when appropriate) are not worth any point.
3. Do put in sufficient explanatory comments in your code.
4. Functions are very important. Do look up the video on the Safari O'Reilly ressource if you are still not clear on them !

For this week you can use these imports at the start of your programs:

In [ ]:
import numpy as np
import matplotlib.pyplot as plt

We will use a new module `uncertainties`, which is **not** standard in the colab environment. You will have to first run:

In [3]:
! pip install -q uncertainties

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.1/60.1 kB 4.2 MB/s eta 0:00:00


To install the module before you can import it:

In [ ]:
import uncertainties as uc
import uncertainties.umath as um # for maths functions

# Introduction
In the practical classes PX2133/PX2233 and PX2338 (Obs tech), as well as your year 3/4 project, a lot of emphasis is placed on the determination and mathematical handling of errors.
The uncertainties module allows us to deal very easily with [error propagation](https://en.wikipedia.org/wiki/Propagation_of_uncertainty). For this sheet you should remind yourself about error bars in measurements and about propagation of uncertainties. Take an example from your lab handbook:

**Example 1**: If the length of a rectangle is $1.24\pm0.02 m$ and its breadth is $0.61\pm0.01 m$, what is its area and the error in the area? The following code snippet solves this problem in a few lines.

In [ ]:
L = uc.ufloat(1.24, 0.02)
W = uc.ufloat(0.61, 0.01)
print ('Area is:', L*W, 'm^2') # Do remember to add the units when printing!

Area is: 0.756+/-0.017 m^2


**Note**: For the area itself, it's fairly straightforward:

In [ ]:
1.24*0.61

0.7564

However, for the error bar on this number:

In [ ]:
0.02*0.01

0.0002

does not work. Instead, the [error progation formula](https://en.wikipedia.org/wiki/Propagation_of_uncertainty#Example_formulae) gives:

In [ ]:
np.abs(1.24*0.61)*np.sqrt((0.02/1.24)**2+(0.01/0.61)**2)

0.01739540169125163

So the area is $0.756\pm0.017 m^2$. `uncertainties` obviously saves a lot of work, even for such a simple case. You can also take a look at the web site uncertainties hosted at https://pythonhosted.org/uncertainties/user_guide.html. In particular, [this section](https://pythonhosted.org/uncertainties/user_guide.html#access-to-the-uncertainty-and-to-the-nominal-value) shows some of the properties of a `ufloat` you can access directly.

**Example 2**: A reference object is $10.0\pm0.0001 m$ long, and makes a viewing angle of $0.62\pm0.02 rad$. How far is it?

In [ ]:
L = uc.ufloat(10.0, 0.0001)
theta = uc.ufloat(0.62,0.02)

Distance = (L/2)/um.tan(theta/2)

print ('Distance is:', Distance.nominal_value, 'm, with an error of:', Distance.std_dev)

Distance is: 15.609024890896208 m, with an error of: 0.537283338762715


Note the need to use "umath" functions (like `um.tan()` instead of `np.tan()`), and how to get the nominal value and the standard deviation of the uncertainties objects. To get nicer looking output, such as controlling the number of significant digits printed, you can use the information about formatting at https://docs.python.org/3/tutorial/inputoutput.html. In the exercises below you need to print the values to the screen. (Don’t forget units.)

# Exercises
This must be marked before you leave the lab. Mark weighting is in brackets.
**Save your work to GitHub after having run all cells with `Runtime` -> `Restart and run all`. And do not change the notebook's filename.** Do add comments to your code, you'll lose points if your code is hard to understand. Graphs without labels (or units when appropriate) are not worth any point.

## Exercise 0
[0] With some approximations, we have measured the mass of the following black-holes:
```
"35.6+/-3.9","30.6+/-3.7","63.1+/-3.2","23.2+/-9.8","13.6+/-4.5","35.7+/-6.8","13.7+/-6.0","7.7+/-2.4","20.5+/-4.0"
```
Compute for each (with error-bars) their lifetime due to Hawking radiation:
$$
t = \left(\frac{M}{M_{\odot}}\right)^3\,\times\,2.097\,\times\,10^{67} yr
$$

(this exercise is for demonstration purposes and won't be marked)

## Exercise 1
[2] An object is measured to travel a distance $x = 5.1 \pm 0.4 m$ during a time of $t = 0.4 \pm 0.1 s$. What is the average velocity and the error in the average velocity?

In [4]:
import uncertainties as uc
x = uc.ufloat(5.1, 0.4)  # distance in meters
t = uc.ufloat(0.4, 0.1)  # time in seconds

velocity = x / t

print('Average velocity is:', velocity, 'm/s')

Average velocity is: 12.7+/-3.3 m/s


## Exercise 2
[2] An enterprising cow attempts to jump over the moon by jumping vertically into the air with initial speed $v_0=4.0\pm0.2 m/s$. After a time $t=0.60\pm0.06s$, the height of the cow is $h = v_0t-\frac{1}{2}g t^2 = 0.636 m$. What is the uncertainty in $h$? Take $g$ as exactly $9.81 ms^{-2}$.

In [5]:
import uncertainties as uc
import uncertainties.umath as um

v0 = uc.ufloat(4.0, 0.2) # initial velocity in m/s
t = uc.ufloat(0.60, 0.06) # time in seconds
g = 9.81 # acceleration due to gravity in m/s^2 (exact value)

h = v0 * t - 0.5 * g * t**2

print('The height of the cow is:', h, 'm')

The height of the cow is: 0.63+/-0.16 m


## Exercise 3
[2] In an optics experiment the object distance $u$ is measured to be 20cm and the image distance $v$ is 10cm, both to an accuracy of 0.5cm. Find the focal length $f$ of the lens using the formula:

$$ \frac{1}{u}+\frac{1}{v}=\frac{1}{f}$$

In [6]:
import uncertainties as uc
import uncertainties.umath as um

u = uc.ufloat(20, 0.5) # object distance in cm
v = uc.ufloat(10, 0.5) # image distance in cm

# Lens formula: 1/f = 1/u + 1/v
f = 1 / (1/u + 1/v)

print('The focal length is:', f, 'cm')

The focal length is: 6.67+/-0.23 cm


## Exercise 4
[2] Two students each measure the refractive index of water. Jack measures a value of $1.33 \pm 0.03$ while Jill measures $1.28 \pm 0.02$. Are these values in agreement? *You do have to think a bit about this one...*

In [7]:
import uncertainties as uc

jack_n = uc.ufloat(1.33, 0.03)
jill_n = uc.ufloat(1.28, 0.02)

difference = jack_n - jill_n

print(f"Jack's measurement: {jack_n}")
print(f"Jill's measurement: {jill_n}")
print(f"Difference: {difference}")

# Check for agreement
# The values are in agreement if the absolute difference between the nominal values
# is less than or equal to the sum of their standard deviations.
if abs(difference.nominal_value) <= jack_n.std_dev + jill_n.std_dev:
    print("The values are in agreement.")
else:
    print("The values are not in agreement.")

Jack's measurement: 1.330+/-0.030
Jill's measurement: 1.280+/-0.020
Difference: 0.05+/-0.04
The values are not in agreement.
The values are not in agreement (alternative check).


## Exercise 5
[2] The damped resonance frequency $\omega_{res}$ of an oscillating system is related to the (un-damped) natural angular frequency $\omega_0$ and the damping coefficient $\alpha$ by:

$$\omega_{res} = \sqrt{ \omega_0^2 - 2\alpha^2}$$

Find $f_0$ if the measured resonance frequency $f_{res}$ is $23.2\pm0.1 Hz$ and the measured damping coefficient is $19.5\pm0.5s^{-1}$.


In [8]:
import uncertainties as uc
import uncertainties.umath as um
import numpy as np

# Given values with uncertainties
f_res = uc.ufloat(23.2, 0.1)  # Damped resonance frequency in Hz
alpha = uc.ufloat(19.5, 0.5) # Damping coefficient in s^-1

# Convert frequency from Hz to angular frequency (omega = 2*pi*f)
omega_res = 2 * np.pi * f_res

# The formula given is omega_res = sqrt(omega_0^2 - 2*alpha^2)
omega_0 = um.sqrt(omega_res**2 + 2 * alpha**2)

# Convert angular frequency back to frequency in Hz (f_0 = omega_0 / (2*pi))
f_0 = omega_0 / (2 * np.pi)

print('The undamped natural frequency f_0 is:', f_0, 'Hz')

The undamped natural frequency f_0 is: 23.61+/-0.10 Hz


## Exercise 6
[2] Suppose you have the following equation from one of your lab experiments:

$$f=\frac{c}{2}\sqrt{\frac{n_x^2}{L_x^2}+\frac{n_y^2}{L_y^2}+\frac{n_z^2}{L_z^2}}$$

where $f$ is the resonant frequency of sound waves in a box of sides $L_x$, $L_y$ and $L_z$ in length and the $n_x$ etc. are integers. $L_x = 10.2\pm0.2m$, $L_y = 5.2\pm0.3m$ and $L_z = 20.0\pm0.1 m$, while $c = 331.3 + T * 0.606 \,m\,s^{-1}$ is the temperature-dependent speed of sound, and the temperature $T$ is $23 \pm 1^\circ C$.
Calculate $f$ and the error in $f$ for the following values of $(nx,ny,nz)= (1,1,1), (1,1,2)$ and $(2,1,1)$.


In [13]:
import uncertainties as uc
import uncertainties.umath as um
import numpy as np

# Given values with uncertainties
Lx = uc.ufloat(10.2, 0.2)  # Length in x-direction in meters
Ly = uc.ufloat(5.2, 0.3)   # Length in y-direction in meters
Lz = uc.ufloat(20.0, 0.1)  # Length in z-direction in meters
T = uc.ufloat(23, 1)      # Temperature in Celsius

# Speed of sound formula
c = 331.3 + T * 0.606  # Speed of sound in m/s

# Resonant frequency formula
def calculate_frequency(nx, ny, nz, c, Lx, Ly, Lz):
  """Calculates the resonant frequency with uncertainties."""
  return (c / 2) * um.sqrt((nx**2 / Lx**2) + (ny**2 / Ly**2) + (nz**2 / Lz**2))

# Calculate and print frequency for each (nx, ny, nz) combination
modes = [(1, 1, 1), (1, 1, 2), (2, 1, 1)]

for nx, ny, nz in modes:
  f = calculate_frequency(nx, ny, nz, c, Lx, Ly, Lz)
  print(f'For (nx, ny, nz) = ({nx}, {ny}, {nz}), the resonant frequency f is: {f:.3f} Hz')

For (nx, ny, nz) = (1, 1, 1), the resonant frequency f is: 38.247+/-1.670 Hz
For (nx, ny, nz) = (1, 1, 2), the resonant frequency f is: 41.065+/-1.556 Hz
For (nx, ny, nz) = (2, 1, 1), the resonant frequency f is: 48.188+/-1.402 Hz


## Exercise 7
[4] The reflection coefficient $R_\parallel$ for parallel plane-polarised light reflected from a surface is given by the equation:

$$ R_\parallel = \frac{\tan^2(\theta_i - \theta_t)}{\tan^2(\theta_i + \theta_t)} $$

Calculate the error in $R_\parallel$ given measurements $\theta_i = (78 \pm 1)^\circ$ and $\theta_t = (40 \pm 1)^\circ$.

In [11]:
import uncertainties as uc
import uncertainties.umath as um
import numpy as np

# Given values with uncertainties in degrees, convert to radians
theta_i_deg = uc.ufloat(78, 1)
theta_t_deg = uc.ufloat(40, 1)

theta_i_rad = um.radians(theta_i_deg)
theta_t_rad = um.radians(theta_t_deg)


# Reflection coefficient formula
R_parallel = (um.tan(theta_i_rad - theta_t_rad))**2 / (um.tan(theta_i_rad + theta_t_rad))**2

print('The reflection coefficient R_parallel is:', R_parallel)

The reflection coefficient R_parallel is: 0.173+/-0.027


## Exercise 8
[4] Calculate and print to the screen the fractional uncertainty, as a percentage to one
significant figure, of the fluid flow discharge coefficient $C_d$ from the equation

$$
C_d = \frac{\dot{m}\sqrt{1-\left(\frac{d}{D}\right)^4}}{Kd^2F\sqrt{\rho\Delta P}}
$$

where

\begin{align*}
    C_d &= \text{discharge coefficient}&& \text{(no units)} \\
    \dot{m} &= \text{mass flow rate}&& = 0.13 \pm 0.01kg\,s^{-1} \\
    d &= \text{orifice diameter}&& = 11\pm 1 mm \\
    D &= \text{pipe diameter}&& = 71 \pm 1 mm \\
    \rho &= \text{fluid density}&& =1.01\pm0.01g\,cm^{-3} \\
    \Delta P &= \text{differential pressure}&& =156 \pm 7 Pa \\
    K &= \text{a constant parameter}&& =\text{constant (no units)} \\
    F &= \text{thermal expansion factor}&& =\text{constant (no units)}
\end{align*}


In [10]:
import uncertainties as uc
import uncertainties.umath as um
import numpy as np

# Given values with uncertainties
m_dot = uc.ufloat(0.13, 0.01) # mass flow rate in kg/s
d = uc.ufloat(11e-3, 1e-3)   # orifice diameter in m (converted from mm)
D = uc.ufloat(71e-3, 1e-3)   # pipe diameter in m (converted from mm)
rho = uc.ufloat(1.01e3, 0.01e3) # fluid density in kg/m^3 (converted from g/cm^3)
Delta_P = uc.ufloat(156, 7)  # differential pressure in Pa

# K and F are constants, so they have no uncertainty.
K = 1
F = 1

# Discharge coefficient formula
Cd = (m_dot * um.sqrt(1 - (d/D)**4)) / (K * d**2 * F * um.sqrt(rho * Delta_P))

# Calculate fractional uncertainty as a percentage
fractional_uncertainty = (Cd.std_dev / abs(Cd.nominal_value)) * 100

print(f"The discharge coefficient Cd is: {Cd}")
print(f"The fractional uncertainty in Cd is: {fractional_uncertainty:.1g}%")

The discharge coefficient Cd is: 2.7+/-0.5
The fractional uncertainty in Cd is: 2e+01%


## Exercise 9: Optional problem (not marked)
If you have time and want to try something interesting, do the following problem by plotting in 2D:
 - Draw an equilateral triangle with vertices and coordinates: vertex 1: $(p_1,q_1)$; vertex 2: $(p_2, q_2)$; vertex 3: $(p_3, q_3)$.
 - Place a dot at an arbitrary point $P = (x_0, y_0)$ within this triangle.
 - Find the next point by selecting randomly an integer $n = 1 , 2, $  or $3$ :
    1. If 1 , place a dot halfway between P and vertex 1.
    2. If 2 , place a dot halfway between P and vertex 2.
    3. If 3 , place a dot halfway between P and vertex 3.
 - Repeat the last two steps using the last dot as the new P.

Mathematically, the coordinates of successive points are given by the formulae

$$(x_{i+1},y_{i+1})=0.5[(x_i,y_i)+(p_n,q_n)]$$

and

$$n=int(1+3r_i),$$

where $r_i$ is a random number between 0 and 1 and where the $int()$ function outputs the closest integer smaller than or equal to the argument.

Try extending this to four vertices.